packages/langium/src/workspace/document-builder.ts

```diff
@@ -133,7 +126,7 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
         affected.forEach(e => {
             const linker = this.serviceRegistry.getServices(e.uri).references.Linker;
             linker.unlink(e);
-            e.state = DocumentState.Indexed;
+            e.state = DocumentState.Processed; // need to re-index potentially linked references
```

Comment:
Be careful not to "upgrade" documents here:
```suggestion
            e.state = Math.min(e.state, DocumentState.Processed); // need to re-index potentially linked references
```